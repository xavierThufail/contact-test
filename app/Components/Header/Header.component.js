import React, { useState } from 'react';
import { View, Text, TextInput, TouchableOpacity, StyleSheet } from 'react-native';

import Icon from '../Icon/Icon.component';
import config from './Header.config';
import styles from './Header.style';
import useHeader from './Header.hook';

const renderFormInput = ({ label, onChangeText, value, last, flex = 1, numeric }) => (
  <View style={styles.containerFlex(last, flex)}>
    <Text style={styles.textBody}>{ label }</Text>
    <TextInput 
      keyboardType={numeric ? 'numeric' : 'default'}
      value={String(value)}
      onChangeText={onChangeText}
      style={styles.input}
    />
  </View>
);

const renderFormGroup = ({ formValue, setFormValue }) => (
  <View>
    <View style={styles.row}>
      { renderFormInput({ label: 'First Name', value: formValue.firstName, onChangeText: firstName => setFormValue({ ...formValue, firstName }) }) }
      { renderFormInput({ label: 'Last Name', value: formValue.lastName, onChangeText: lastName => setFormValue({ ...formValue, lastName }), last: true }) }
    </View>
    <View style={styles.row}>
      { renderFormInput({ label: 'Age', value: formValue.age, onChangeText: age => setFormValue({ ...formValue, age }), flex: 1, numeric: true }) }
      { renderFormInput({ label: 'Link Photo', value: formValue.photo, onChangeText: photo => setFormValue({ ...formValue, photo }), last: true, flex: 3 }) }
    </View>
  </View>
);

const renderButton = ({ color, onPress, title }) => (
  <TouchableOpacity
    style={styles.button(color)}
    onPress={onPress}
    activeOpacity={1}
  >
    <Text style={styles.buttonText}>{ title }</Text>
  </TouchableOpacity>
);

const renderForm = ({ formValue, setFormValue, handleAddContact }) => (
  <View style={{
    display: 'flex',
    paddingHorizontal: 10
  }}>
    { renderFormGroup({ formValue, setFormValue }) }
    <View style={styles.row}>
      <View style={styles.containerFlex(true, 1)}>
        { renderButton({
          title: 'Add Contact',
          onPress: handleAddContact
        }) }
      </View>
    </View>
  </View>
);

const renderHeader = ({ handleOpenClose, showForm, title }) => (
  <View style={{...styles.container, zIndex: 100}}>
    {title && <Text style={styles.title}>{title}</Text>}

    <View>
      { renderButton({
        color: showForm ? '#eb2323': '',
        title: showForm ? 'Cancel' : 'Add Contact',
        onPress: handleOpenClose
      }) }
    </View>
  </View>
);

const Header = ({title}) => {
  const { formValue, setFormValue, showForm, handleOpenClose, handleAddContact } = useHeader();

  return (
    <View>
      { renderHeader({ handleOpenClose, showForm, title }) }
      { showForm && renderForm({ setFormValue, formValue, handleAddContact }) }
    </View>
  )
};

Header.propTypes = config.propTypes;
Header.displayName = config.displayName;

export default Header;
