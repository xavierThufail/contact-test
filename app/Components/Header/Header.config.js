import PropTypes from 'prop-types';

const propTypes = {
  title: PropTypes.string.isRequired,
};

const displayName = 'Header';

export default {
  propTypes,
  displayName
};
