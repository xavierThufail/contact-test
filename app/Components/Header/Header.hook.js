import React, { useState } from "react";
import { useDispatch } from 'react-redux';

import allAction from "../../Store/actions";

const useHeader = () => {
  const dispatch = useDispatch();

  const [showForm, setShowForm] = useState(false);
  const [formValue, setFormValue] = useState({
    firstName: '',
    lastName: '',
    age: '',
    photo: ''
  });

  const handleAddContact = () => {
    dispatch(allAction.contact.postContact(formValue));
    handleOpenClose();
  };

  const handleOpenClose = () => {
    setFormValue({
      firstName: '',
      lastName: '',
      age: '',
      photo: ''
    })
    setShowForm(!showForm)
  };

  return {
    showForm,
    formValue,
    setFormValue,
    handleOpenClose,
    handleAddContact
  }
};

export default useHeader;
