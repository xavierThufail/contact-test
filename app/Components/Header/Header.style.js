const styles = {
  container: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    marginTop: 10,
    paddingHorizontal: 10,
  },
  title: {
    fontSize: 22,
    fontFamily: 'source-sans',
    color: "#747474"
  },
  button: (color) => ({
    minHeight: 30,
    minWidth: 50,
    borderRadius: 10,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: color ? color : '#1b83e3',
    padding: 10,
  }),
  buttonText: {
    textAlign: 'center',
    color: '#f0f0f0',
    fontWeight: 'bold'
  },
  containerFlex: (last, flex) => ({
    flex,
    marginRight: last ? 0 : 10
  }),
  textBody: {
    fontWeight: 'bold'
  },
  row: {
    flexDirection: 'row',
    alignItems: 'center', 
  },
  input: {
    borderWidth: 1,
    height: 40,
    borderRadius: 10,
    marginBottom: 10,
    paddingLeft: 20
  }
};

export default styles;
