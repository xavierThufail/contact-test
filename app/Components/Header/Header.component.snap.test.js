import React from 'react';
import { render } from '@testing-library/react-native';
import { Provider } from 'react-redux';

import Header from './Header.component';
import store from '../../Store';


jest
  .mock('@react-navigation/native', () => ({
    useNavigation: jest.fn(() => ({
      navigate: jest.fn()
    }))
  }))
  .mock('../Icon', () => 'Icon')

describe('Header Component snapshot test', () => {
  it('Should render correctly', () => {
    const mockedProps = {
      title: 'Contact'
    };

    const { toJSON } = render(
      <Provider store={store}>
        <Header {...mockedProps} />
      </Provider>
    );

    expect(toJSON()).toMatchSnapshot()
  })
});
