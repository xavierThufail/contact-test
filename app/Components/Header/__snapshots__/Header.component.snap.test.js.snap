// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component snapshot test Should render correctly 1`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginTop": 10,
        "paddingHorizontal": 10,
        "zIndex": 100,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#747474",
          "fontFamily": "source-sans",
          "fontSize": 22,
        }
      }
    >
      Contact
    </Text>
    <View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#1b83e3",
            "borderRadius": 10,
            "justifyContent": "center",
            "minHeight": 30,
            "minWidth": 50,
            "opacity": 1,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#f0f0f0",
              "fontWeight": "bold",
              "textAlign": "center",
            }
          }
        >
          Add Contact
        </Text>
      </View>
    </View>
  </View>
</View>
`;
